File: app/src/main/java/jgappsandgames/smartreminderslite/holder/CheckpointHolder.java
package jgappsandgames.smartreminderslite.holder;

// Android OS
import android.content.Context;
import android.content.Intent;

// Views
import android.graphics.Paint;
import android.os.Vibrator;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.View.OnLongClickListener;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.CompoundButton.OnCheckedChangeListener;
import android.widget.TextView;

// Program
import jgappsandgames.smartreminderslite.R;

import jgappsandgames.smartreminderslite.tasks.TaskActivity;
import jgappsandgames.smartreminderslite.tasks.checkpoint.CheckpointActivity;
import jgappsandgames.smartreminderslite.utility.ActivityUtility;

// Save
import jgappsandgames.smartreminderssave.tasks.Checkpoint;

/**
 * CheckpointHolder
 * Created by joshua on 9/1/17.
 */
public class CheckpointHolder implements OnClickListener, OnLongClickListener, OnCheckedChangeListener {
    // Data
    private final Checkpoint checkpoint;
    private final TaskActivity activity;
    private final String task;

    // Views
    private final TextView text;
    private final Button edit;

    // Initializer
    public CheckpointHolder(TaskActivity activity, String task, Checkpoint checkpoint, View view) {
        super();

        // Set Data
        this.activity = activity;
        this.task = task;
        this.checkpoint = checkpoint;

        // Find Views
        text = view.findViewById(R.id.text);
        edit = view.findViewById(R.id.edit);

        // Set Click Listeners
        text.setOnClickListener(this);
        text.setOnLongClickListener(this);
        edit.setOnClickListener(this);
        edit.setOnLongClickListener(this);

        // Set Views
        setViews();
    }

    // View Handler
    private void setViews() {
        if (checkpoint.status) {
            text.setPaintFlags(text.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
        } else {
            text.setPaintFlags(text.getPaintFlags() & (~Paint.STRIKE_THRU_TEXT_FLAG));
        }
        text.setText(checkpoint.text);
    }

    // Click Handlers
    @Override
    public void onClick(View view) {
        if (view.equals(edit)) {
            Intent intent = new Intent(activity, CheckpointActivity.class);

            intent.putExtra(ActivityUtility.CHECKPOINT, checkpoint.toString());
            intent.putExtra(ActivityUtility.TASK_NAME, task);

            activity.startActivityForResult(intent, ActivityUtility.REQUEST_CHECKPOINT);
        } else if (view.equals(text)) {
            checkpoint.status = !checkpoint.status;
            setViews();
            activity.editCheckpoint(checkpoint);
        }
    }

    @Override
    public boolean onLongClick(View view) {
        activity.deleteCheckpoint(checkpoint);

        Vibrator v = (Vibrator) activity.getSystemService(Context.VIBRATOR_SERVICE);
        try {
            if (v != null && v.hasVibrator()) v.vibrate(100);
        } catch (NullPointerException n) {
            n.printStackTrace();
        }

        return true;
    }

    // Check Handler
    @Override
    public void onCheckedChanged(CompoundButton view, boolean status) {
        checkpoint.status = status;
        activity.editCheckpoint(checkpoint);
    }
}

File: app/src/main/java/jgappsandgames/smartreminderslite/holder/CheckpointHolder.kt
package jgappsandgames.smartreminderslite.holder

// Android OS
import android.content.Context
import android.content.Intent
import android.graphics.Paint
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator

// Views
import android.view.View
import android.widget.Button
import android.widget.TextView

// App
import jgappsandgames.smartreminderslite.R
import jgappsandgames.smartreminderslite.tasks.TaskActivity
import jgappsandgames.smartreminderslite.tasks.CheckpointActivity
import jgappsandgames.smartreminderslite.utility.CHECKPOINT
import jgappsandgames.smartreminderslite.utility.REQUEST_CHECKPOINT
import jgappsandgames.smartreminderslite.utility.TASK_NAME

// Save
import jgappsandgames.smartreminderssave.tasks.Checkpoint

/**
 * CheckpointHolder
 * Created by joshua on 12/30/2017.
 *
 * Handles the View For Drawing the Checkpoints in the Task
 */
class CheckpointHolder(private val activity: TaskActivity, private val task: String, private val checkpoint: Checkpoint, view: View):
        View.OnClickListener, View.OnLongClickListener {
    // Views ---------------------------------------------------------------------------------------
    private val text: TextView = view.findViewById(R.id.text)
    private val edit: Button = view.findViewById(R.id.edit)

    // Constructor ---------------------------------------------------------------------------------
    init {
        // Set Click Listeners
        text.setOnClickListener(this)
        text.setOnLongClickListener(this)
        edit.setOnClickListener(this)
        edit.setOnLongClickListener(this)

        // Set Views
        setViews()
    }

    // View Handler --------------------------------------------------------------------------------
    private fun setViews() {
        if (checkpoint.status) text.paintFlags = text.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
        else text.paintFlags = text.paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
        text.text = checkpoint.text
    }

    // Click Handlers ------------------------------------------------------------------------------
    override fun onClick(view: View) {
        if (view == edit) {
            val intent = Intent(activity, CheckpointActivity::class.java)
            intent.putExtra(CHECKPOINT, checkpoint.toString())
            intent.putExtra(TASK_NAME, task)
            activity.startActivityForResult(intent, REQUEST_CHECKPOINT)
        } else if (view == text) {
            checkpoint.status = !checkpoint.status
            setViews()
            activity.editCheckpoint(checkpoint)
        }
    }

    override fun onLongClick(view: View): Boolean {
        activity.deleteCheckpoint(checkpoint)

        val v = activity.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator?
        try {
            if (v != null && v.hasVibrator()) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    v.vibrate(VibrationEffect.createOneShot(100, VibrationEffect.DEFAULT_AMPLITUDE))
                } else {
                    @Suppress("DEPRECATION")
                    v.vibrate(100)
                }
            }
        } catch (n: NullPointerException) {
            n.printStackTrace()
        }

        return true
    }
}